File: algorithms/dynamic-programming/travelling-salesman/kotlin/TravellingSalesman.kt
fun travellingSalesman(arr: IntArray): Int {
    val n = arr[0]
    if (n <= 1) return 0
    val dist = Array(n) { i -> IntArray(n) { j -> arr[1 + i * n + j] } }
    val INF = Int.MAX_VALUE / 2
    val full = (1 shl n) - 1
    val dp = Array(1 shl n) { IntArray(n) { INF } }
    dp[1][0] = 0
    for (mask in 1..full) for (i in 0 until n) {
        if (dp[mask][i] >= INF || mask and (1 shl i) == 0) continue
        for (j in 0 until n) {
            if (mask and (1 shl j) != 0) continue
            val nm = mask or (1 shl j)
            val cost = dp[mask][i] + dist[i][j]
            if (cost < dp[nm][j]) dp[nm][j] = cost
        }
    }
    var result = INF
    for (i in 0 until n) result = minOf(result, dp[full][i] + dist[i][0])
    return result
}

File: algorithms/dynamic-programming/travelling-salesman/cpp/travelling_salesman.cpp
#include <vector>
#include <algorithm>
#include <climits>

int travelling_salesman(std::vector<int> arr) {
    int n = arr[0];
    if (n <= 1) return 0;
    std::vector<std::vector<int>> dist(n, std::vector<int>(n));
    for (int i = 0; i < n; i++)
        for (int j = 0; j < n; j++)
            dist[i][j] = arr[1 + i*n + j];

    int INF = INT_MAX / 2;
    int full = (1 << n) - 1;
    std::vector<std::vector<int>> dp(1 << n, std::vector<int>(n, INF));
    dp[1][0] = 0;

    for (int mask = 1; mask <= full; mask++)
        for (int i = 0; i < n; i++) {
            if (dp[mask][i] >= INF || !(mask & (1 << i))) continue;
            for (int j = 0; j < n; j++) {
                if (mask & (1 << j)) continue;
                int nm = mask | (1 << j);
                dp[nm][j] = std::min(dp[nm][j], dp[mask][i] + dist[i][j]);
            }
        }

    int result = INF;
    for (int i = 0; i < n; i++)
        result = std::min(result, dp[full][i] + dist[i][0]);
    return result;
}
